sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
+                                                          .error_message = AZ_SPAN_NULL,
+                                                          .error_tracking_id = AZ_SPAN_NULL,
+                                                          .error_timestamp = AZ_SPAN_NULL };
+}
+
+/*
+Documented at
+https://docs.microsoft.com/en-us/rest/api/iot-dps/runtimeregistration/registerdevice#deviceregistrationresult
+  "registrationState":{
+    "x509":{},
+    "registrationId":"paho-sample-device1",
+    "createdDateTimeUtc":"2020-04-10T03:11:13.0276997Z",
+    "assignedHub":"contoso.azure-devices.net",
+    "deviceId":"paho-sample-device1",
+    "status":"assigned",
+    "substatus":"initialAssignment",
+    "lastUpdatedDateTimeUtc":"2020-04-10T03:11:13.2096201Z",
+    "etag":"IjYxMDA4ZDQ2LTAwMDAtMDEwMC0wMDAwLTVlOGZlM2QxMDAwMCI="}}
+*/
+AZ_INLINE az_result az_iot_provisioning_client_payload_error_code_parse(
+    az_json_token_member* tm,
+    az_iot_provisioning_client_registration_state* out_state)
+{
+  if (az_span_is_content_equal(AZ_SPAN_FROM_STR("errorCode"), tm->name))
+  {
+    double value;
+    AZ_RETURN_IF_FAILED(az_json_token_get_number(&tm->token, &value));
+    out_state->extended_error_code = (uint32_t)value;
+    out_state->status = (uint32_t)value / 1000;
+
+    return AZ_OK;
+  }
+
+  return AZ_ERROR_ITEM_NOT_FOUND;
+}
+
+AZ_INLINE az_result az_iot_provisioning_client_payload_registration_state_parse(
+    az_json_parser* jp,
+    az_json_token_member* tm,
+    az_iot_provisioning_client_registration_state* out_state)
+{
+  if (tm->token.kind != AZ_JSON_TOKEN_OBJECT_START)
+  {
+    return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
+  }
+
+  bool found_assigned_hub = false;
+  bool found_device_id = false;
+
+  while ((!(found_device_id && found_assigned_hub))
+         && az_succeeded(az_json_parser_parse_token_member(jp, tm)))
+  {
+    if (az_span_is_content_equal(AZ_SPAN_FROM_STR("assignedHub"), tm->name))
+    {
+      found_assigned_hub = true;
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm->token, &out_state->assigned_hub_hostname));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("deviceId"), tm->name))
+    {
+      found_device_id = true;
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm->token, &out_state->device_id));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("errorMessage"), tm->name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm->token, &out_state->error_message));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("lastUpdatedDateTimeUtc"), tm->name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(&tm->token, &out_state->error_timestamp));
+    }
+    else if (tm->token.kind == AZ_JSON_TOKEN_OBJECT_START)
+    {
+      AZ_RETURN_IF_FAILED(az_json_parser_skip_children(jp, tm->token));
+    }
+    else
+    {
+      (void)az_iot_provisioning_client_payload_error_code_parse(tm, out_state);
+    }
+  }
+
+  return AZ_OK;
```

Comment:
should we return error if (found_assigned_hub && found_device_id) != true?

Reply:
We want the chain-of-responsibility semantic: https://github.com/Azure/azure-sdk-for-c/blob/master/sdk/iot/doc/mqtt_state_machine.md#receiving-apis
If the topic was meant for this particular API then we want it to return OK if we were able to parse the payload.

There are several levels of error reporting: API-level, request-level and operation-level. To avoid any confusion, the function returns errors only on API-level (cannot parse, protocol errors, etc).
If the topic/payload are correct, we return OK and require the application to check the request-level or operation level errors through the returned structure.